### PR TITLE
Add pagination and sorting mechanism

### LINKED ISSUE #50: Allow slicing of data messages
Today, the SDMX Rest ws standard foresees one way for restricting number of observations through the firstNObservations and lastNObservations URL parameters. But those apply individually to each "series" and not to the whole data message, and are thus not useful when the client needs to restrict the message size, or for chunking the request into smaller pieces that are better manageable by the client. 

A nice solution would be provided by the HTTP protocol that includes a way to request only a specific **range of data**. The related [RFC 7233](https://tools.ietf.org/html/rfc7233#page-8) standard says: 

> The "Range" header field on a GET request modifies the method semantics to request transfer of only one or more subranges of the selected representation data, rather than the entire selected representation data.

It works in the following way: 

1. The server indicates with the HTTP header field "**Accept-Ranges**" that it supports range requests for the target resource, and which types of range units, e.g. bytes.
`Accept-Ranges: bytes` 

2. The client uses the "**Range**" header field in the GET request to indicate the requested subrange, e.g. for the first 1000 bytes (with 0-based indexes):
`Range: bytes=0-999`
And for the next 1000 bytes:
`Range: bytes=1000-1999`

3. The server, if the range request is admissible, returns the **206 (Partial Content)** status code that indicates that the server is successfully fulfilling a range request for the target resource by transferring one or more parts of the selected representation that correspond to the satisfiable ranges found in the request's "**Range**" header field. It also returns a "**Content-Range**" header field, describing what range of the selected representation is enclosed, e.g.
`Content-Range: bytes 0-999/1234`
Note: 1234 is the total number of bytes for the complete resource, or  
`Content-Range: bytes 0-999/*`
Note: * is indicating that the total number of bytes for the complete resource is unknown by the server, or 
`Content-Range: bytes 0-566/567` 
Alternatively, the server generates a **416 (Range Not Satisfiable)** response and a "**Content-Range**" header field with an unsatisfied-range value "*", e.g.
`Content-Range: bytes */1234 `

Also see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) and [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range), for examples of requests for multiple sub-ranges.

## However

Since the `bytes` unit is inappropriate for our purposes, in .Stat Suite we have tested the HTTP `Range` header approach with another range unit: `values`.  While this works fine in most environments, some users have reported issues in AWS hosting scenarios. Indeed, some cloud hosting services do not allow using range units that have not been registered with IANA, 

We therefore decided to use a non-standard/proprietary `X-Range` HTTP header instead of `Range`. This worked in all environments. The specification we have used is as follow:

1. The server indicates with the HTTP header field "**Accept-Ranges**" that it supports range requests for the target resource and the range unit `values`.:
`Accept-Ranges: values` 

2. The client uses the "**X-Range**" header field in the GET request to indicate the subrange of the selected representation data, e.g. for the first 1000 SDMX observation values:
`X-Range: values=0-999`
And for the next 1000 values:
`X-Range: values=1000-1999`

3. The server, if the range request is admissible, returns the **206 (Partial Content)** status code that indicates that the server is successfully fulfilling a range request for the target resource by transferring one or more parts of the selected representation that correspond to the satisfiable ranges found in the request's "**X-Range**" header field. It also returns a "**Content-Range**" header field, describing what range of the selected representation is enclosed, e.g.
`Content-Range: values 0-999/1234`
Note: 1234 is the total number of values for the complete resource, or  
`Content-Range: values 0-999/*`
Note: * is indicating that the total number of values for the complete resource is unknown by the server, or 
`Content-Range: values 0-566/567` 
Alternatively, the server generates a **416 (Range Not Satisfiable)** response and a "**Content-Range**" header field with an unsatisfied-range value "*", e.g.
`Content-Range: values */1234 `

### Note

This ticket should be address together with ticket [Allow ordering the data query results by dimensions](https://github.com/sdmx-twg/sdmx-rest/issues/151), since slicing/pagination makes only really sense if the order of the observations is deterministic.

### LINKED ISSUE #151: Allow ordering the data query results by dimensions
Linked to ticket https://github.com/sdmx-twg/sdmx-rest/issues/50, which we use to retrieve only partial data that fit into a size-limited, paginated, online pivot table, the need is for a **user-defined order of the retrieved observations/series**. Since the tables are pivot tables, the required content (observations) of the first (and subsequent) table pages depend on the table layout (distribution of dimensions on different tables axes, like rows and columns). The table layout thus dictates which slices of observations need to be returned per page. Therefor, the SDMX web services call should be able to accept a user-defined sequence of dimensions, which is used to order the observations before they are returned.

The order within one dimension should take the code order (respectively the hierarchical code order) into account.

### PR DESCRIPTION
**Pull request by sosna**
Fix #50 
Fix #151

### CODE REVIEW

**Review by agent96 — DISMISSED** [review]

**Comment by dosse** [comment]
@sosna I made the following changes:

Rest-cheat-sheet:
- syntax corrections for `e.g.`

Pagination:
- added series key for serieskeyonly queries

Sort:  
- by default, always group observations into time series and sort by ascending time period
- within components sort by code ID or non-coded value

**Something that still needs clarification is the following:**

If the data are sorted by the components in the order stated in the sort clause, then `sort=TIME_PERIOD:desc` would do the following: It would first sort in descending order of TIME_PERIOD and then by the other components in a nondeterministic way, e.g.,

```
  FREQ,DIM1,TIME_PERIOD,OBS_VALUE
  A,A,2022,1.0654
  A,B,2022,1.0261
  A,A,2021,1.0752
  A,B,2021,1.0730
```

In consequence, the observations will not be grouped by time series. I don't think that this was intended with `sort=TIME_PERIOD:desc`, right?

This could be solved by saying:   
The data are sorted first in a nondeterministic way for any component not included in the sort parameter and then in the order of components in the sort parameter. In this way, `sort=TIME_PERIOD:asc` would also be the default sorting value if the data has a time dimension.

Ok for you?

**Review by stratosn — APPROVED** [review]

**Comment by dosse** [comment]
@sosna @stratosn @agent96 I have updated the pull request with the changes discussed and agreed this Feb per email, notably adding the feature to use `*` in the sort as replacement for all components (incl. TIME_PERIOD). For me, the specs are fully ready. We can have this reviewed and approved by TWG in June meeting next week if ok for you.

File: doc/data.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ measures | String | This parameter specifies the measures to be returned. Possib
...
+sort | String | This parameter specifies the order in which the returned data should be sorted. It contains either one or more component IDs, by which the data should be sorted, separated by `+` (to indicate an AND), the `*` operator, which represents all dimensions as positioned in the DSD, or the keyword `series_key`, which represents all non-time dimensions as positioned in the DSD. The sorting must respect the sequence, in which the components are listed. In addition, each component, or the set of components (through the operator or keyword) can be sorted in ascending or descending order by appending `:asc` or `:desc`, with `:asc` being the default. For any component not included in the sort parameter, the related order is non-deterministic. Except for time periods, which have a natural chronological order, the sorting within a component is based on the code IDs or the non-coded component values. | | No
```
Comment (sosna) [inline doc/data.md:30]: We seem to consider timeseries data only, and thereby exclude cross-sectional data. That might be OK (not sure) but, if we want to make it more generic, we could rephrase "or the keyword `series_key`, which represents all non-time dimensions as positioned in the DSD" as follow:

"or the keyword `series_key`, which represents all dimensions, as positioned in the DSD, except the dimension at the observation level (TIME_PERIOD for timeseries data, or any dimension except TIME_PERIOD for cross-sectional data)"

File: api/sdmx-rest.yaml [hunk 2]
```diff
@@ -87,156 +97,160 @@ paths:
...
+        - $ref: "#/components/parameters/providers"
+        - $ref: "#/components/parameters/resourceIDs"
+        - $ref: "#/components/parameters/versions"
+        - $ref: "#/components/parameters/metaDetail"
+        - $ref: "#/components/parameters/accept-encoding"
+        - $ref: "#/components/parameters/accept-language"
+        - $ref: "#/components/parameters/if-modified-since"
       responses:
         <<: *common_responses
+<<<<<<< cr-50_151
```
Comment (sosna) [inline api/sdmx-rest.yaml:250]: Same as above, i.e. conflict must be resolved.

File: api/sdmx-rest.yaml [hunk 3]
```diff
@@ -447,8 +461,13 @@ components:
     providers:
       in: path
       name: providerID
+<<<<<<< cr-50_151
```
Comment (sosna) [inline api/sdmx-rest.yaml:464]: Same as above (conflict resolution)

File: api/sdmx-rest.yaml [hunk 4]
```diff
@@ -1,82 +1,92 @@
 openapi: 3.0.0
 info:
+<<<<<<< cr-50_151
```
Comment (sosna) [inline api/sdmx-rest.yaml:3]: Conflict must be resolved, else we won't be able to merge.

**Comment by sosna** [comment]
Thanks, @dosse. The PR looks good to me!

File: api/sdmx-rest.yaml [hunk 5]
```diff
@@ -357,7 +359,7 @@ components:
             metadataproviderscheme,
             reportingtaxonomy,
             metadataprovisionagreement
```
Comment (stratosn) [inline api/sdmx-rest.yaml:361]: Is a comma missing after `metadataprovisionagreement`?

File: api/sdmx-rest.yaml [hunk 6]
```diff
@@ -618,10 +620,48 @@ components:
...
+      schema:
+        type: integer
+        minimum: 1
+    sort:
+      in: query
+      name: sort
+      description: |
+        This parameter specifies the order in which the returned data should be sorted. 
+
+        It contains either one or more component IDs, by which the data should be sorted, separated by `+` (to indicate an AND), the `*` operator, which represents all dimensions as positioned in the DSD, or the keyword `series_key`, which represents all non-time dimensions as positioned in the DSD.
```
Comment (stratosn) [inline api/sdmx-rest.yaml:648]: Do we not foresee querying for data with another Dimension at obs? In such a case, the `series_key` would need to cover that, as well.

Comment (sosna) [reply api/sdmx-rest.yaml]: Thanks, @stratosn. Addressed in https://github.com/sdmx-twg/sdmx-rest/pull/185/commits/ad7476a8e1e8f75fe46832b1ffe5e7ef8b1c0baa

File: api/sdmx-rest.yaml (hunk 5 above)
Comment (sosna) [reply api/sdmx-rest.yaml]: Indeed, thanks @stratosn. Fixed in https://github.com/sdmx-twg/sdmx-rest/pull/185/commits/1a682f490096da1b9c59dd3f4cca15e13ad3b1ed

**Review by stratosn — APPROVED** [review]

**Pull request merged**